### PR TITLE
Fix ArduSub autotest

### PR DESCRIPTION
**Pull request by Williangalvani**
These changes fix the autotest for ArduSub
There are three main changes:

- Don't wait for "tilt alignment complete" message, <s>as Subs are not calibrated after reboot by default</s>. It is not sent consistently with sitl and causing the tests to fail.

- Don't test arming/disarming with switches (it doesn't support switches)

- Don't test arming/disarming with RC inputs (it also doesn't support 'stick commands')

### CODE REVIEW

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -196,7 +196,10 @@ def fetch_parameters(self):
     def reboot_sitl(self):
         """Reboot SITL instance and wait it to reconnect."""
         self.mavproxy.send("reboot\n")
-        self.mavproxy.expect("tilt alignment complete")
+        self.mav.wait_heartbeat() # makes sure we have a mav_type
+        # ArduSub doesn't do tilt alignment after boot
+        if self.mav.mav_type != mavutil.mavlink.MAV_TYPE_SUBMARINE:
```
Comment (khancyr) [inline Tools/autotest/common.py]: intersting ! I am pretty sure it was working yesterday when I test it ....

**Comment by khancyr** [comment]
Hello,
Thanks for the contribution.
 I already got the fixes for SUB here : https://github.com/ArduPilot/ardupilot/pull/9393

File: Tools/autotest/common.py (hunk 1 above)
Comment (khancyr) [reply Tools/autotest/common.py]: ok it doesn't work on my patch, so it was surely working I rebase ..

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -196,7 +196,10 @@ def fetch_parameters(self):
     def reboot_sitl(self):
         """Reboot SITL instance and wait it to reconnect."""
         self.mavproxy.send("reboot\n")
-        self.mavproxy.expect("tilt alignment complete")
+        self.mav.wait_heartbeat() # makes sure we have a mav_type
```
Comment (khancyr) [inline Tools/autotest/common.py]: I don't think this is autotest proof ... we could still get a heartbeat just after issuing a reboot to mavproxy.

File: Tools/autotest/common.py [hunk 3]
```diff
@@ -196,7 +196,10 @@ def fetch_parameters(self):
     def reboot_sitl(self):
         """Reboot SITL instance and wait it to reconnect."""
         self.mavproxy.send("reboot\n")
-        self.mavproxy.expect("tilt alignment complete")
```
Comment (khancyr) [inline Tools/autotest/common.py]: this is still send on SUB

File: Tools/autotest/common.py (hunk 1 above)
Comment (khancyr) [reply Tools/autotest/common.py]: the reboot is working fine on master, what error did you get ?

**Comment by khancyr** [comment]
I extract the fix for sub from my pr into https://github.com/ArduPilot/ardupilot/pull/9550, they should into master soon.
I did more tests and the reboot is working fine...

File: Tools/autotest/common.py (hunk 1 above)
Comment (Williangalvani) [reply Tools/autotest/common.py]: I get this testing ArduSub on master:
```
....
Init ArduSub V3.6-dev (820f9bd2)

Free RAM: 131072
load_all took 0us
0 0 0 validate_structures:348: Validating structures
DataFlash_File: buffer size=16384

Init completeReceived 850 parameters
Saved 833 parameters to test.ArduSub/logs/2018-10-12/flight1/mav.parm
APM: EKF2 IMU0 Origin set to GPS
APM: EKF2 IMU1 Origin set to GPS
bind port 5762 for 2
Serial port 2 on TCP port 5762
bind port 5763 for 3
Serial port 3 on TCP port 5763
TSYS01 device is null!APM: MYGCS: 255, heartbeat lost
APM: EKF2 IMU1 is using GPS
APM: EKF2 IMU0 is using GPS
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
APM: MYGCS: 255, heartbeat lost
Flight battery warning
APM: MYGCS: 255, heartbeat lost
Timed out looking for tilt alignment complete
...
```

File: Tools/autotest/common.py (hunk 3 above)
Comment (jaxxzer) [reply Tools/autotest/common.py]: Is it sent when INS_GYR_CAL is set to 0 (our default)?

File: Tools/autotest/common.py (hunk 1 above)
Comment (OXINARF) [reply Tools/autotest/common.py]: @Williangalvani That log doesn't show anything from SITL after the reboot, so either it didn't reboot  or it took too long to reboot. Neither of which means the message isn't there. That message comes from the EKF, I don't see how it wouldn't be present on Sub.

File: Tools/autotest/common.py (hunk 3 above)
Comment (Williangalvani) [reply Tools/autotest/common.py]: It is not sending here, I'll try to find out why

**Comment by Williangalvani** [comment]
Can you test it? your pr failed for me both in waiting for the "tilt alignment" message and a heading timeout later (maybe changed flight mode? I'll check)

File: Tools/autotest/common.py (hunk 3 above)
Comment (khancyr) [reply Tools/autotest/common.py]: @jaxxzer, I don't know. But that is not the default param that we get on SITL.
Currently it is send. I agree that the caching of the text that we currently have isn't great...
I will be quite busy on the next two weeks but I will make a good pass on Sub simulation and autotest after... it is quite behind the others...

File: Tools/autotest/common.py (hunk 1 above)
Comment (khancyr) [reply Tools/autotest/common.py]: From my tests I saw this sometime, the error comes generaly from the expect that isn't capable to catch the right text.
@oxinarf, the message from sitl are here! the instance reboot well but the expect thing failed... we need to change that

File: Tools/autotest/common.py (hunk 3 above)
Comment (Williangalvani) [reply Tools/autotest/common.py]: It seems to be a parameters issue, I'm taking a further look

**Comment by Williangalvani** [comment]
So, I found that changing the parameters so that SR0_EXTRA3 is not zero solves the issue of the "tilt alignment complete" not arriving, diff updated.

**Comment by peterbarker** [comment]
On Thu, 11 Oct 2018, Pierre Kancir wrote:

> I already got the fixes for SUB here : #9393

Those patches are much more extensive than these ones.

If I merge this one (#9546), can you rebase on top of it when you're 
getting back to this ins a couple of weeks?

**Comment by peterbarker** [comment]
On Fri, 12 Oct 2018, Willian Galvani wrote:

> So, I found that changing the parameters so that SR0_EXTRA3 is not zero
> solves the issue of the "tilt alignment complete" not arriving, diff
> updated.

What you may find here is that it's actually a lack of heartbeats from the 
GCS which is causing the problem - thus the mygcs error message :-)

The problem is that MAVProxy is sending heartbeats on wall-clock time, but 
ArduSub is running at 10x speedup.  Whether a heartbeat from MAVProxy 
arrives while waiting for the tilt message is very much rolling the dice.

Setting the SR0_EXTRA3 parameter has the effect of starting the tilt 
messages streaming, so a heartbeat isn't required.  It's quite a 
work-around.

Now tridge has some magic in the offing to do with timing in 
SITL/mavproxy.  I'm inclined to pull this workaround in on the basis that 
it gets things working, with a note to try to come back and fix it when 
we've got this new stuff.

**Comment by jaxxzer** [comment]
> The problem is that MAVProxy is sending heartbeats on wall-clock time, but
ArduSub is running at 10x speedup.  Whether a heartbeat from MAVProxy
arrives while waiting for the tilt message is very much rolling the dice.

AHA! xD

I'm looking this over myself now.

**Comment by OXINARF** [comment]
Not sure it helps anyone but... https://github.com/OXINARF/MAVProxy/commits/pr/support-sitl-speedup

**Comment by Williangalvani** [comment]
I changed reboot_sitl() to wait for "Initialising APM" instead of  "tilt alignment complete", this seems to work with all vehicles without requiring parameters changes.

File: Tools/autotest/common.py [hunk 4]
```diff
@@ -1318,16 +1318,20 @@ def test_arm_feature(self):
...
-        self.set_rc(arming_switch, 1000)
-        if not self.arm_motors_with_switch(arming_switch):
-            raise NotAchievedException()
-        if not self.disarm_motors_with_switch(arming_switch):
-            raise NotAchievedException()
-        self.set_rc(arming_switch, 1000)
+        # Rover and Sub don't have auto disarm, Sub has no 'switches'
+        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_SUBMARINE:
+            if not self.disarm_vehicle():
+                self.progress("Failed to DISARM")
```
Comment (jaxxzer) [inline Tools/autotest/common.py]: match the other vehicles here.
```suggestion
                raise NotAchievedException()
```

File: Tools/autotest/common.py [hunk 5]
```diff
@@ -1318,16 +1318,20 @@ def test_arm_feature(self):
...
-        self.start_test("Test arm and disarm with switch")
-        arming_switch = 7
-        self.set_parameter("RC%d_OPTION" % arming_switch, 41)
-        self.set_rc(arming_switch, 1000)
-        if not self.arm_motors_with_switch(arming_switch):
-            raise NotAchievedException()
-        if not self.disarm_motors_with_switch(arming_switch):
-            raise NotAchievedException()
-        self.set_rc(arming_switch, 1000)
+        # Rover and Sub don't have auto disarm, Sub has no 'switches'
```
Comment (jaxxzer) [inline Tools/autotest/common.py]: The 'auto disarm' comment should move up to the relevant block of code.

File: Tools/autotest/common.py [hunk 6]
```diff
@@ -1352,27 +1358,31 @@ def test_arm_feature(self):
...
-        self.set_parameter("ARMING_RUDDER", 1)
-        self.arm_vehicle()
-        if self.disarm_motors_with_rc_input():
-            self.progress("Failed to NOT ARM")
-            raise NotAchievedException()
-        self.disarm_vehicle()
-        self.mav.wait_heartbeat()
-        self.set_parameter("ARMING_RUDDER", 2)
+
+        # Sub Sub doesn't have 'stick commands'
```
Comment (jaxxzer) [inline Tools/autotest/common.py]: `Sub Sub Sub`

**Comment by jaxxzer** [comment]
> I changed reboot_sitl() to wait for "Initialising APM" instead of "tilt alignment complete", this seems to work with all vehicles without requiring parameters changes.

I think this is fine for what we have here, a minimal patch to get autotest moving again for all vehicles.

We can discuss further other methods to detect reboot like system time when @peterbarker and @khancyr have some more time to look and input on this.

**Comment by peterbarker** [comment]
On Wed, 17 Oct 2018, Willian Galvani wrote:

> I changed reboot_sitl() to wait for "Initialising APM" instead of "tilt alignment complete", this seems to work with all vehicles without requiring
> parameters changes.

There are two problems with using "Initialising APM"
  - we want to change it to not say "APM" at some stage :-)
  - it's actually only emitted as a side-effect of the main loop going out 
to lunch for a while, not because we're in boot-up!

Perhaps we should just look at the board's boot time and base our decision 
on that number wrapping.  That would be looking at the output of 
`get_sim_time()`.

**Comment by Williangalvani** [comment]
I agree that is how it should be done, but it lead me to another issue: 
[Time is not wrapping](https://pastebin.com/k3GM1fQN), the first SYSTEM_TIME message received after the reboot is has ta timestamp higher than before rebooting.
[It seems to be more consistent without speedup=10](https://pastebin.com/PdmbedeB).

**Comment by jaxxzer** [comment]
> There are two problems with using "Initialising APM"
  - we want to change it to not say "APM" at some stage :-)

"Initialising APM" is just as arbitrary as "tilt alignment complete" and this demonstrates the problem with using statustext for this, this is not maintainable. No change here except that Initialising APM actually works here where tilt alignment complete does not.

>  - it's actually only emitted as a side-effect of the main loop going out
to lunch for a while, not because we're in boot-up!

And my understanding is that the current "tilt alignment complete" is a side effect of the ekf starting, not because we're in boot up. So no change here.



> Perhaps we should just look at the board's boot time and base our decision
on that number wrapping.  That would be looking at the output of
`get_sim_time()`.

Agreed, I've already proposed this. I think that much warrants a new issue/pr. This PR fixes autotest for all vehicles with minimal patch, we need another one to improve the methods.

**Comment by OXINARF** [comment]
> "Initialising APM" is just as arbitrary as "tilt alignment complete" and this demonstrates the problem with using statustext for this, this is not maintainable. No change here except that Initialising APM actually works here where tilt alignment complete does not.

@jaxxzer It's not the same for a simple reason: the Initialising message comes from each vehicle, so if one changes, the autotest will need a special case, while the til alignment one is in the EKF and if it is changed, it's just one change in autotest.

In any case there is something that doesn't make sense to me: both messages are STATUSTEXT so why does one appears and the other apparently doesn't? And why only on reboot?

**Comment by jaxxzer** [comment]
@jaxxzer It's not the same for a simple reason: the Initialising message comes from each vehicle, so if one changes, the autotest will need a special case, while the til alignment one is in the EKF and if it is changed, it's just one change in autotest.

It's the same that in the event the message changes, it will need to be updated in more than one place or the test will break. That's unmaintainable and I maintain my point here. Are there other differences? Sure.

> In any case there is something that doesn't make sense to me: both messages are STATUSTEXT so why does one appears and the other apparently doesn't? And why only on reboot?

I don't know. I've gotten completely inconsistent/unsatisfying results running sitl at different speedups. The output should be the same regardless of speedup, but alas, that is not the case.

**Comment by OXINARF** [comment]
> It's the same that in the event the message changes, it will need to be updated in more than one place or the test will break.

No, that's not true, with the tilt message it's one place, just like it's changing in this PR now.

**Comment by jaxxzer** [comment]
one place vs two places or three, its still more than it needs to be/should be.

**Comment by peterbarker** [comment]
On Thu, 18 Oct 2018, Willian Galvani wrote:

> I agree that is how it should be done, but it lead me to another issue:
> Time is not wrapping, the first SYSTEM_TIME message received after the reboot is has ta timestamp higher than before rebooting.
> It seems to be more consistent without speedup=10.

This might be why I went with the tilted message thing initially.

But this shouldn't really be a problem - we should just be looping waiting 
for the time to wrap.

**Comment by jaxxzer** [comment]
> But this shouldn't really be a problem - we should just be looping waiting
for the time to wrap.

It shouldn't be, but it is. The sitl behavior is inconsistent with each run. Sometimes the first system time is sent 20s after boot (first run), sometimes 80s after boot (second run), thus when the sitl instance is  booted, parameters changed, and immediately rebooted, a time wrap is not detected.

Thus my push to get this patch in AS IS. It solves a specific problem, and **it works** as well and better than everything that has been committed to date.

Other problems and improvements regarding sitl and autotest have been presented in the discussion here, and I think those should be addressed in another patch as they go beyond the scope of this pull request.

**Comment by jaxxzer** [comment]
> No, that's not true, with the tilt message it's one place, just like it's changing in this PR now.

@OXINARF It's more than one place every time and everywhere that you use statustext. It's in the vehicle code or whatever code sends the statustext, and it's also in the autotest or whatever code looks for the statustext.

My point is this: In my opinion, statustext is an asanine way of determining the vehicle state programatically. We should be using state indicators in the vehicle, sensor, subsystem etc. health and status flags in the mavlink stream.

File: Tools/autotest/common.py (hunk 3 above)
Comment (jaxxzer) [reply Tools/autotest/common.py]: From what I've gathered testing this, the "tilt alignment complete" message should be sent regardless of parameters, or at least the ones mentioned so far right now.

This message not appearing during certain boots seems to have something to do with running with `--speedup 10`. It appears consistently with speedup 1 (default).

**Comment by peterbarker** [comment]
Anything (pretty much) to get things actually functional again.

Apologies for the breakage in the first place.

**Pull request merged**